stdlib/universe/holt_winters.go

```diff
@@ -0,0 +1,463 @@
+	// Building schema.
+	if err := execute.AddTableKeyCols(tbl.Key(), builder); err != nil {
+		return err
+	}
+	newTimeIdx, err := builder.AddCol(flux.ColMeta{
+		Label: execute.DefaultTimeColLabel,
+		Type:  flux.TTime,
+	})
+	if err != nil {
+		return err
+	}
+	newValueIdx, err := builder.AddCol(flux.ColMeta{
+		Label: execute.DefaultValueColLabel,
+		Type:  flux.TFloat,
+	})
+	if err != nil {
+		return err
+	}
+
+	// Cleaning data for HoltWinters input.
+	vs, start, stop, err := hwt.getCleanData(tbl, colIdx, timeIdx)
+	if err != nil {
+		return err
+	}
+
+	// Holt Winters.
+	hw := holt_winters.New(int(hwt.n), int(hwt.s), hwt.withFit)
+	bf := func() *array.Float64Builder {
+		return array.NewFloat64Builder(fluxarrow.NewAllocator(hwt.alloc))
+	}
+	newVs := hw.Do(vs, bf)
+	// don't need vs anymore
+	vs.Release()
+
+	// Crafting timestamps.
+	// Timestamps are deduced by summing the interval to the first/last valid timestamp.
+	tsb := array.NewInt64Builder(fluxarrow.NewAllocator(hwt.alloc))
+	s := stop + int64(hwt.interval)
+	if hwt.withFit {
+		s = start
+	}
+	for i := 0; i < newVs.Len(); i++ {
+		tsb.Append(s)
+		s += int64(hwt.interval)
+	}
+	newTs := tsb.NewInt64Array()
+	defer func() {
+		newVs.Release()
+		newTs.Release()
+	}()
+
+	// Appending columns.
+	if err := builder.AppendTimes(newTimeIdx, newTs); err != nil {
+		return err
+	}
+	if err := builder.AppendFloats(newValueIdx, newVs); err != nil {
+		return err
+	}
+	for j, c := range tbl.Key().Cols() {
+		v := tbl.Key().Value(j)
+		idx := execute.ColIdx(c.Label, builder.Cols())
+		if err := hwt.craftKeyCols(builder, idx, v, newVs.Len()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// getCleanData returns cleaned data (using the value and time column), and the first and last valid timestamps.
+// Below are the cleaning criteria.
+// Rows that have a null timestamp get discarded.
+// Rows that have a null value are considered invalid, but used by the algorithm.
+// HoltWinters supposes to work with evenly spaced values in time, so:
+//  - the Interval passed to the transformation is used to divide the data in time buckets;
+//  - if many values are in the same bucket, the first one is selected, the others are skipped;
+//  - if no value is present for a bucket, that is considered as an invalid value (treated like null values).
+// HoltWinters will only be provided with the values returned.
+// Timestamps can be deduced by summing interval to the first/last valid timestamp.
+func (hwt *holtWintersTransformation) getCleanData(tbl flux.Table, colIdx, timeIdx int) (*array.Float64, int64, int64, error) {
+	vs := array.NewFloat64Builder(fluxarrow.NewAllocator(hwt.alloc))
+	var start, stop int64
+	bucketEnd := int64(-1)
+	bucketFilled := false
+	roundTime := func(t int64) int64 {
+		interval := int64(hwt.interval)
+		remainder := t % interval
+		if remainder > interval/2 {
+			// round up
+			return (t/interval + 1) * interval
+		}
+		// round down
+		return (t / interval) * interval
```

Comment:
return t?

Reply:
Sorry, copy/paste from InfluxQL. Definitely  `return t`!

Reply:
Mmmmm 🤔 No, it's different `/`  is the integer division. The code is correct.